genesis/engine/entities/fem_entity.py

```diff
@@ -750,6 +751,113 @@ def set_muscle_direction(self, muscle_direction):
             muscle_direction=muscle_direction,
         )
 
+    def set_vertex_constraints(
+        self, vertex_indices, target_positions=None, link=None, constraint_type="hard", stiffness=0.0, damping=0.0
+    ):
+        """
+        Set vertex constraints for specified vertices.
+
+        Parameters
+        ----------
+            vertex_indices : array_like
+                List of vertex indices to constrain.
+            target_positions : array_like, shape (len(vertex_indices), 3), optional
+                List of target positions [x, y, z] for each vertex. If not provided, the initial positions are used.
+            link : RigidLink
+                Optional rigid link for the vertices to follow, maintaining relative position.
+            constraint_type: str
+                A "hard" constraint directly set position and zero velocity.
+                A "soft" constraint uses a spring force to pull the vertex towards the target position.
+            stiffness : float
+                Specify a spring stiffness for a soft constraint.
+        """
+        if self.solver._use_implicit_solver:
+            gs.logger.warning("Ignoring vertex constraint; unsupported with FEM implicit solver.")
+            return
+
+        if not self.solver._constraints_initialized:
+            self.solver.init_constraints()
+
+        assert constraint_type in ["hard", "soft"], "constraint_type must be 'hard' or 'soft'."
+        constraint_type_int = 0 if constraint_type == "hard" else 1
+        vertex_indices = torch.as_tensor(vertex_indices, dtype=gs.tc_int, device=gs.device) + self._v_start
+        self._assert_vertex_idx_in_bounds(vertex_indices)
+
+        if target_positions is None:
+            target_positions = self.init_positions[vertex_indices]
+        target_positions = torch.as_tensor(target_positions, dtype=gs.tc_float, device=gs.device)
+
+        if link is None:
+            link_idx = -1
+            link_init_pos = np.zeros((3,))
+            link_init_quat = np.zeros((4,))
+        else:
+            assert hasattr(link, "get_pos") and hasattr(
+                link, "get_quat"
+            ), "Link must have get_pos() and get_quat() methods, e.g. RigidLink"
+            link_init_pos = link.get_pos().detach().cpu().numpy()
```

Comment:
`link_init_pos = tensor_to_array(link.get_pos())`

Could you extend `tensor_to_array` to also take care of `.detach()` ? I don't think it does this already.